scripts/diagnostics/afn_auditor.py

```diff
@@ -0,0 +1,400 @@
+        # print(large_external_links, too_many_external_links, way_too_many_external_links)
+
+        #
+        # Machine-readable output
+        #
+        self.json['multizone_link_histogram'] = link_histogram
+        self.json['max_multizone_links'] = {
+            'zone': self.nodes[max_link_node_name]['zone_name'],
+            'afn_zone': max_link_node_name,
+            'count': max_links
+        }
+        self.json['external_link_histogram'] = external_link_histogram
+        self.json['max_external_links'] = {
+            'zone': self.nodes[max_external_link_node_name]['zone_name'],
+            'afn_zone': max_external_link_node_name,
+            'count': max_external_links
+        }
+        self.json['messages'] = []
+        if large_links > 0:
+            mesg = '%d zone(s) with greater than 25 links' % large_links
+            if too_many_links > 0:
+                mesg += ', %d with greater than 50 links' % too_many_links
+                if way_too_many_links > 0:
+                    mesg += ', %d with greater than 100 links' % way_too_many_links
+                    mesg += ', model performance may suffer'
+            self.json['messages'].append(mesg)
+        if large_external_links > 0:
+            mesg = '%d zone(s) with greater than 8 external links' % large_external_links
+            if too_many_external_links > 0:
+                mesg += ', %d with greater than 16 external links' % too_many_external_links
+                if way_too_many_external_links > 0:
+                    mesg += ', %d with greater than 32 external links' % way_too_many_external_links
+                    mesg += ', model performance may suffer'
+            self.json['messages'].append(mesg)
+    
+
+if __name__ == '__main__':
+    #
+    # The main body of the script, do argument processing first
+    #
+    parser = argparse.ArgumentParser(description='AirflowNetwork model audit script')
+    # args.add_argument('-g', '--graph', help='Generate a graphviz graph',
+    #                  default=False, action='store_true')
+    parser.add_argument('-g', '--graph', help='generate a graphviz .dot output file',
+                        dest='graph', metavar='dotfile', default=None,
+                        type=argparse.FileType('w'))
+    parser.add_argument('-p', '--pretty', help='write pretty JSON output',
+                        default=False, action='store_true')
+    parser.add_argument("json_file")
+
+    args = parser.parse_args()
+
+    fp = open(args.json_file, 'r')
+    my_model = json.load(fp)
+    fp.close()
+
+    auditor = AFNAuditor(my_model)
+
+    auditor.audit()
+
+    #
+    # Now write it all out
+    #
+    indent = None
+    if args.pretty:
+        indent = 2
+
+    json.dump(auditor.json, sys.stdout, indent=indent)
+
+    if args.graph:
+        #
+        # Generate a graph
+        #
+        # Give nodes names for displaying
+        this_count = 0
+        for _, this_node in external_nodes.items():
```

Comment:
`external_nodes`, `nodes`, and `surfs` are all undefined here, are these supposed to be member variables on an instance of the auditor?

Reply:
@Myoldmopar There's an issue right now with different code on different computers that I'll resolve this morning (and move this code all to the member function rather than this code.